src/com/amazon/ion/impl/_Private_IonReaderBuilder.java

```diff
@@ -0,0 +1,172 @@
+            }
+            if (IonStreamUtils.isIonBinary(ionData, offset, length)) {
+                return makeIncrementalReader(this, new ByteArrayInputStream(ionData, offset, length));
+            }
+        }
+        return makeReader(validateCatalog(), ionData, offset, length, lstFactory);
+    }
+
+    /**
+     * Determines whether a stream that begins with the bytes in the provided buffer could be binary Ion.
+     * @param buffer up to the first four bytes in a stream.
+     * @param length the actual number of bytes in the buffer.
+     * @return true if the first 'length' bytes in 'buffer' match the first 'length' bytes in the binary IVM.
+     */
+    private static boolean startsWithIvm(byte[] buffer, int length) {
+        for (int i = 0; i < length; i++) {
+            if (_Private_IonConstants.BINARY_VERSION_MARKER_1_0[i] != buffer[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public IonReader build(InputStream ionData)
+    {
+        InputStream wrapper = ionData;
+        if (isIncrementalReadingEnabled()) {
+            if (!ionData.markSupported()) {
+                wrapper = new BufferedInputStream(ionData);
+            }
+            wrapper.mark(_Private_IonConstants.BINARY_VERSION_MARKER_SIZE);
+            byte[] possibleIVM = new byte[_Private_IonConstants.BINARY_VERSION_MARKER_SIZE];
+            int bytesRead;
+            try {
+                bytesRead = wrapper.read(possibleIVM);
+                wrapper.reset();
+            } catch (IOException e) {
+                throw new IonException(e);
+            }
+            if (IonStreamUtils.isGzip(possibleIVM, 0, possibleIVM.length)) {
+                throw new IllegalArgumentException("Automatic GZIP detection is not supported with incremental" +
+                    "support enabled. Wrap the bytes with a GZIPInputStream and call build(InputStream).");
+            }
+            // If the input stream is growing, it is possible that fewer than BINARY_VERSION_MARKER_SIZE bytes are
+            // available yet. Simply check whether the stream *could* contain binary Ion based on the available bytes.
+            // If it can't, fall back to text.
+            // NOTE: if incremental text reading is added, there will need to be logic that handles the case where
+            // the reader is created with 0 bytes available, as it is impossible to determine text vs. binary without
+            // reading at least one byte. Currently, in that case, just create a binary incremental reader. Either the
+            // stream will always be empty (in which case it doesn't matter whether a text or binary reader is used)
+            // or it's a binary stream (in which case the correct reader was created) or it's a growing text stream
+            // (which has always been unsupported).
+            if (startsWithIvm(possibleIVM, bytesRead)) {
```

Comment:
I changed this. Previously it required the entire IVM to match, meaning that you couldn't create an incremental reader over a stream that initially contains less than 4 bytes.